PR auto-merge improvements

cristiand391 opened the pull request:
This PR improves the current way `gh` handles an auto-merge by making it schedule an auto-merge **only** when the `mergeStateStatus` field is `BLOCKED` (otherwise it should just merge it).

Ref. https://github.com/cli/cli/issues/3514#issuecomment-828379835

Closes #3514 
Fixes #3858

cristiand391 committed:
Only schedule an auto-merge when PR state is blocked

When passing `--auto` flag, only schedule an auto-merge if
the `mergeStateStatus` field is "BLOCKED".
This ensures that a PR will always be merged when passing `--auto` even
if it doesn't have required checks or if checks have already passed.

cristiand391 committed:
add test case

mislav commented on pkg/cmd/pr/merge/merge.go:
```diff
@@ -202,12 +202,13 @@ func mergeRun(opts *MergeOptions) error {
-			setCommitBody: opts.BodySet,
+			repo:             baseRepo,
+			pullRequestID:    pr.ID,
+			mergeStateStatus: pr.MergeStateStatus,
```
Instead of adding an additional field to `mergePayload`, could the caller figure out whether `auto` is necessary or not based on the value of `pr.MergeStateStatus`? I feel like this is more of a responsibility of the caller than the responsibility of the `mergePullRequest()` function.

mislav commented on pkg/cmd/pr/merge/merge.go:
```diff
@@ -156,7 +156,7 @@ func mergeRun(opts *MergeOptions) error {
 	findOptions := shared.FindOptions{
 		Selector: opts.SelectorArg,
-		Fields:   []string{"id", "number", "state", "title", "lastCommit", "mergeable", "headRepositoryOwner", "headRefName"},
+		Fields:   []string{"id", "number", "state", "title", "lastCommit", "mergeable", "mergeStateStatus", "headRepositoryOwner", "headRefName"},
```
It's worth noting that adding this field to the query most likely makes the `gh pr merge` command unusable for GitHub Enterprise versions older than 2.20, since the field was only added then. Those GHE versions are not officially supported anymore, so I guess this is fine.

mislav reviewed (changes requested):
The approach looks good! Left a suggestion about avoiding having to do changes to `mergePullRequest()` function and its payload

mislav commented:
A related feature idea that you could consider adding @cristiand391 (no worries if not): if a regular `gh pr merge` (no `--auto`) fails because PR status is BLOCKED, perhaps the error message can suggest that you redo the invocation and add `--auto` to have the PR being merged as soon as it's unblocked.

cristiand391 committed:
Simplify auto-merge detection

cristiand391 committed:
Suggest to enable auto-merge when PR merge state is BLOCKED

cristiand391 commented:
> if a regular gh pr merge (no --auto) fails because PR status is BLOCKED, perhaps the error message can suggest that you redo the invocation and add --auto to have the PR being merged as soon as it's unblocked.

Done! 🤘🏾 

![automerge-error](https://user-images.githubusercontent.com/6853656/122646703-0ba55a80-d110-11eb-9ebe-8acfa9b4210c.png)

samcoe commented on pkg/cmd/pr/merge/merge.go:
```diff
@@ -196,16 +196,25 @@ func mergeRun(opts *MergeOptions) error {
+
 	deleteBranch := opts.DeleteBranch
 	crossRepoPR := pr.HeadRepositoryOwner.Login != baseRepo.RepoOwner()
+	autoMerge := opts.AutoMergeEnable && pr.MergeStateStatus == "BLOCKED"
```
I think this conditional is not restrictive enough based on the potential values of `MergeStateStatus`. According to https://docs.github.com/en/graphql/reference/enums#mergestatestatus I feel like we should only try to actually merge the PR when the  `MergeStateStatus` is one of `UNSTABLE `, `HAS_HOOKS`, or `CLEAN`. Having a whitelist approach also protects us from any updates in new values are added to `MergeStateStatus`.

cristiand391 replied:
Sorry for the late reply, the only thing the `MergeStateStatus` field is used for is to determine if a PR is in the "BLOCKED" state (in which case it enables auto-merge).

Currently if you try to merge a PR that has some conflict (either a merge conflict or checks failing), gh warns you that there's a conflict:
https://github.com/cli/cli/blob/33c3fb5cdd110f08791bf6afbdad522c414e7a54/pkg/cmd/pr/merge/merge.go#L193-L198

Do you think there should be an additional check based only on `MergeStateStatus`? If so, what benefits would it provide over the current one relying on the `Mergeable` field?

samcoe replied:
Sorry for my confusing comment above. I think I could have better illustrated it with code.

```golang
	approved := pr.MergeStateStatus == "CLEAN" || pr.MergeStateStatus == "HAS_HOOKS" || pr.MergeStateStatus == "UNSTABLE"
	if !opts.AutoMergeEnable && !approved {
		fmt.Fprintf(
			opts.IO.ErrOut,
			"%s Merging is blocked. Enable auto-merge to automatically merge it when all requirements are met with: gh pr merge %d --auto\n",
			cs.FailureIcon(), pr.Number)
		return cmdutil.SilentError
	}

	deleteBranch := opts.DeleteBranch
	crossRepoPR := pr.HeadRepositoryOwner.Login != baseRepo.RepoOwner()
	autoMerge := opts.AutoMergeEnable && !approved
```
The `Mergeable` field, as you stated, is already checked on a previous line but that only checks that `git` can actually merge the PR without conflicts. The `MergeStateStatus` checks that branch checks have also passed, both of these checks are necessary to merge a PR. Values of `CLEAN`, `HAS_HOOKS`, and `UNSTABLE` are the values of `MergeStateStatus` that mean the branch checks have passed. So rather than than say `MergeStateStatus` != `BLOCKED` I would rather check that `MergeStateStatus` is equal to one of the passing values.

cristiand391 committed:
Only merge PRs if branch checks have passed

cristiand391 committed:
Fix tests

samcoe reviewed (approved):
This looks good to me, thanks for making the requested changes! 🙇

mislav committed:
Merge remote-tracking branch 'origin' into improve-automerge

mislav committed:
Adjust conditions for switching between regular and auto merge

Conditions prohibiting a regular merge: BLOCKED, BEHIND, DIRTY.

Conditions triggering a regular merge even if `--auto` was set: CLEAN,
HAS_HOOKS.

Note that UNKNOWN status does not trigger either of the conditions.

mislav reviewed (approved):
Thank you! I've adjusted the feature so that it's more conservative: especially when it comes to UNKNOWN status which is fairly common in my testing.

mislav committed:
Add "UNSTABLE" to immediately mergeable statuses

This status describes a state where the head branch is mergeable and
technically not blocked per base branch requirements, but it does have
non-passing checks.

mislav committed:
Never prompt to delete branch if `--auto` was given
